main.cpp

```diff
@@ -164,84 +175,249 @@ good to go!
+    FloatType& multiply(const IntType& rhs);
+    FloatType& divide(const IntType& rhs);
+
 };
 struct DoubleType
 {
-    double add(double lhs, double rhs);
-    double subtract(double lhs, double rhs);
-    double multiply(double lhs, double rhs);
-    double divide(double lhs, double rhs);
+    double* value;
+    double value_;
+    DoubleType(double value_) : value(new double(value_)){}
+    ~DoubleType()
+    {
+        delete value;
+        value = nullptr;
+    }
+    DoubleType& add(double rhs);
+    DoubleType& subtract(double rhs);
+    DoubleType& multiply(double rhs);
+    DoubleType& divide(double rhs);
+
+    DoubleType& add(const DoubleType& rhs);
+    DoubleType& subtract(const DoubleType& rhs);
+    DoubleType& multiply(const DoubleType& rhs);
+    DoubleType& divide(const DoubleType& rhs);
+
+    DoubleType& add(const FloatType& rhs);
+    DoubleType& subtract(const FloatType& rhs);
+    DoubleType& multiply(const FloatType& rhs);
+    DoubleType& divide(const FloatType& rhs);
+
+    DoubleType& add(const IntType& rhs);
+    DoubleType& subtract(const IntType& rhs);
+    DoubleType& multiply(const IntType& rhs);
+    DoubleType& divide(const IntType& rhs);
 };
 struct IntType
 {
-    int add(int lhs, int rhs);
-    int subtract(int lhs, int rhs);
-    int multiply(int lhs, int rhs);
-    int divide(int lhs, int rhs);
+    int* value;
+    int value_;
+    IntType(double value_) : value(new int(value_)){}
+    ~IntType()
+    {
+        delete value;
+        value = nullptr;
+    }
+    IntType&  add(int rhs);
+    IntType& subtract(int rhs);
+    IntType& multiply(int rhs);
+    IntType& divide(int rhs);
+
+    IntType&  add(const IntType& rhs);
+    IntType& subtract(const IntType& rhs);
+    IntType& multiply(const IntType& rhs);
+    IntType& divide(const IntType& rhs);
+
+    IntType&  add(const FloatType& rhs);
+    IntType& subtract(const FloatType& rhs);
+    IntType& multiply(const FloatType& rhs);
+    IntType& divide(const FloatType& rhs);
+
+    IntType&  add(const DoubleType& rhs);
+    IntType& subtract(const DoubleType& rhs);
+    IntType& multiply(const DoubleType& rhs);
+    IntType& divide(const DoubleType& rhs);
 };
 
-float FloatType::add(float lhs, float rhs)
+FloatType& FloatType::add(float rhs)
 {
-    return lhs + rhs;
+    value += rhs;
+    return *this;
 }
 
-float FloatType::subtract(float lhs, float rhs)
+FloatType& FloatType::subtract(float rhs)
 {
-    return lhs + rhs;
+    value -= rhs;
+    return *this;
 }
 
-float FloatType::multiply(float lhs, float rhs)
+FloatType& FloatType::multiply(float rhs)
 {
-    return lhs * rhs;
+    value *= rhs;
+    return *this;
 }
 
-float FloatType::divide(float lhs, float rhs)
+FloatType& FloatType::divide(float rhs)
 {
-    return lhs / rhs;
+    if (rhs == 0)
+    {
+        std::cout << "dividing by 0 is not allowed in this type ! ";
+        return 0;
+    }
+    else
+    {
+        value /= rhs;
+        return *this;
+    }
```

Comment:
You never returned 0 before when dividing floats.   Why are you suddenly doing it now?
- [ ] revise